---
title: "giant_virus_virsorter"
author: "Caroline Martin"
date: "6/29/2021"
output: html_document
---
# load tidyverse for visualization tabular data
```{r}
library(tidyverse)
```
# load our Virsorter data
```{r}
all_viral_scores <- read_csv("all_viral_scores_binID_gtdb.csv")
```
## Contigs filtered with at least 0.8 NCLDV viral score, as well as at least 10 kb, and also do not have a pre-annotated gtdb taxa
```{r}
gv8_contigs <- all_viral_scores %>%
  filter(NCLDV > 0.8) %>% #the bin has a contig with a GV score above 0.8
  filter(length > 10000) %>%  #contig length over 10kb 
  filter(gtdb_taxa == "None")#contig has no prior gtdb taxa annotation
```

```{r}
gv8_contigs
```
# load DT to create interactive datatable for vusualization of filtered contigs
```{r}
library(DT)
datatable(gv8_contigs)
```
#Further filter these contigs for only contigs  no previous gaint virus annotation, summarize the number of hallmark genes present, summarize the viral numerical score, summarize cellular score, only show binID, NCLDV score, and the IMG link corresponding. Only show contigs with at least one halmark gene present, and a cellular score less than 10 (possible contamination). 
```{r}
gv8_contigs %>% 
  filter(max_score_group == "NCLDV") %>% 
  group_by(contigID) %>% filter(is.na(`giant virus`)) %>% summarize(Hallmark = sum(hallmark),
            Viral = sum(viral), Cellular = sum(cellular), binID, NCLDV, `IMG link`) %>% 
  filter(Hallmark > 0) %>% 
  filter(Cellular < 10)
```

## Consolidate bins with contigs of interest
```{r}
gv8_contigs %>% 
  filter(max_score_group == "NCLDV") %>% filter(!is.na(`giant virus`)) %>% 
  group_by(binID) %>% 
  summarize(Hallmark = sum(hallmark), Viral = sum(viral), Cellular = mean(cellular), `giant virus`) %>% 
  filter(Hallmark > 0) 
```

## Separate Bins of interest separately 
```{r}
gv8_contigs %>% 
  filter(max_score_group == "NCLDV") %>% 
  filter(binID == "3300020659-4") %>% 
  rename(giant_virus = "giant virus") %>% 
  summarize(giant_virus, binID, contigID, Hallmark = sum(hallmark), Viral = sum(viral), Cellular = sum(cellular), binID, NCLDV) %>% 
  filter(Hallmark > 0) %>% 
  filter(Cellular < 10)
```

```{r}
gv8_contigs %>% 
  filter(max_score_group == "NCLDV") %>% 
  filter(binID == "3300020668-3") %>% 
  summarize(`giant virus`, contigID, Hallmark = sum(hallmark), Viral = sum(viral), Cellular = sum(cellular), binID, NCLDV, `IMG link`) 
  
```

```{r}
gv8_contigs %>% 
  filter(max_score_group == "NCLDV") %>% 
  filter(binID == "3300020905-1") %>% 
  rename(giant_virus = "giant virus") %>% 
  summarize(giant_virus, contigID, Hallmark = sum(hallmark), Viral = sum(viral), Cellular = sum(cellular), binID, NCLDV) %>% 
  filter(Hallmark > 0) #%>% 
  #filter(Cellular < 10)
```

```{r}
all_viral_scores %>%  filter(binID == "3300020626-2")
```
# Filters the bins that are annotated as a giant virus 

#Summarize length of annotated GV
```{r}
all_viral_scores %>% group_by(binID, `giant virus`) %>% 
  filter(!is.na(`giant virus`)) %>% 
  group_by(binID) %>% 
  summarize(length = sum(length), Viral = mean(viral), Cellular = mean(cellular), `giant virus`)
```
# NCLDV Score of Known NCLDV from Schulz 2020

```{r}
all_viral_scores %>% group_by(binID == '3300020744-1') %>% 
  filter(!is.na(`giant virus`)) %>% 
  group_by(binID) %>% 
  summarize(Hallmark = sum(hallmark), Viral = sum(viral), `giant virus`)
Terestrivirus
```

#Create tables of individual previously annotated giant viruses (from first study in 2017 Hiddendiversity paper)

```{r}
Hyperion <- all_viral_scores %>%  filter(binID == "3300020757-5")
```

```{r}
Sylvan <- all_viral_scores %>%  filter(binID == "3300020974-4")
```

```{r}
Homa <- all_viral_scores %>%  filter(binID == "3300020904-8")
```

```{r}
Daso <- all_viral_scores %>%  filter(binID == "3300020831-4")
```

```{r}
Solum <- all_viral_scores %>%  filter(binID == "3300020823-3")
```

```{r}
Barre <- all_viral_scores %>%  filter(binID == "3300020795-4	")
```

```{r}
Satyr <- all_viral_scores %>%  filter(binID == "3300020767-5")
```

```{r}
Harvfo <- all_viral_scores %>%  filter(binID == "3300020765-1")
```

```{r}
Gaea <- all_viral_scores %>%  filter(binID == "3300020744-1")
```

```{r}
Soli <- all_viral_scores %>%  filter(binID == "3300020659-4")
```

```{r}
Edafo <- all_viral_scores %>%  filter(binID == "3300020651-1")
```

```{r}
terrestri <- all_viral_scores %>%  filter(binID == "3300020631-1")
```
# Combine all known NCLDVS into one table called Known_NCLDV
```{r}
Known_NCLDV <- rbind(Hyperion, Sylvan, Homa, Daso, Solum, Barre, Satyr, Harvfo, Gaea, Soli, Edafo, terrestri)
Known_NCLDV
```
# Summarize length and mean viral score of all known NCLDV from first study (which can be found in Schulz 2020 data as well)

```{r}
Known_NCLDV %>%  group_by(MinimetagenomeID) %>% 
  summarize(length = sum(length), Viral = mean(viral), `giant virus`, NCLDV = mean(NCLDV))
```

```{r}
library(ggplot2)
```

```{r}
# convert MinimetagenomeID by factor
#Make a histogram of gv8_contigs by minimetagenomeID
# Count greater than 10, who are they
# was minimetagenome and contig the same 
Hist_gv8_miniID <- ggplot(data = gv8_contigs, aes(gv8_contigs$MinimetagenomeID)) + geom_histogram(col = 'blue', fill = 'purple') + labs(title="Histogram of MinimetagenomeID for filtered Giant Virus contigs", x="MinimetagenomeID", y="Count")
Hist_gv8_miniID
```
```{r}
#Make a histogram of all viral scores by minimetagenomeID
Hist_all_miniID <- ggplot(data = all_viral_scores, aes(all_viral_scores$binID)) + geom_histogram(col = 'yellow', fill = 'pink') + labs(title = "Histogram of minimetagenomeID for all viral scores", x = "binID", y = "Count")
Hist_all_miniID
```
```{r}
all_viral_scores_NCLDV <- all_viral_scores %>% group_by(binID) %>% filter(max_score_group == 'NCLDV')
all_viral_scores_NCLDV
```

```{r}
# Histogram of summary scores for gv8_contigs  
Hist_summary_scores_gv8 <- ggplot(data = gv8_contigs, aes(gv8_contigs$max_score)) + geom_histogram(col = 'gray', fill = 'maroon') + labs(title = "Histrogram of summary scores in filtered Giant Virus contigs", x = "Summary Scores (max score) for NCLDV", y = "Count")
Hist_summary_scores_gv8
```

```{r}
# Histogram of summary scores in all Viral Scores data for all max score group as NCLDV
Hist_summary_scores_all_NCLDV <- ggplot(data = all_viral_scores_NCLDV, aes(all_viral_scores_NCLDV$max_score)) + geom_histogram(col = 'maroon', fill = 'gray') + labs(title = 'Histogram of summary scores for all NCLDV in all viral scores', x = "Summary Scores for NCLDV", y = "Count")
Hist_summary_scores_all_NCLDV
```

```{r}
#How many contigs within each binID for gv8_contigs? 
gv8_contigs_per_binID <- group_by(binID) %>% summarize()
gv8_contigs_per_binID
```
# Histogram of contig lengths
