---
title: "Lab 10"
author: "Andrea Dame"
date: "11/2/2020"
output: html_document
---

```{r}
library(vcfR)
```

```{r}
data(vcfR_example)
```


#The meta region contains information about the file, its creation, as well as information to interpret abbreviations used elsewhere in the file
```{r}
strwrap(vcf@meta[1:7])
```
# To summarize this meta data to be easier to read.
```{r}
queryMETA(vcf)
```

```{r}
queryMETA(vcf, element = 'DP')
```
# To show only the acronym for just Format
```{r}
queryMETA(vcf, element = 'FORMAT=<ID=DP')
```

```{r}
head(getFIX(vcf))
```



```{r}
vcf <- read.vcfR("myVCFdata_filtered.vcf.gz")
```

```{r}
head(vcf)
```

```{r}
write.vcf(vcf, "myVCFdata_filtered(1).vcf.gz")
```

#Exercises Part I
# 1. How would we find more information about read.vcfR()?

# You can learn more about a package or function in package by putting a ? in front of it.
```{r}
?read.vcfR()
```

#2. How would we learn what the acronym “AD” stands for?
# Acronyms are explained in the metadata section of a VCF. They can be summarized or searched using query.
```{r}
queryMETA(vcf, element = 'AD')
```

# 3. We used the head() function to view the first few lines of fix data. How would we view the last few lines of fix data?

# We can use the tail function to see the last few lines of fixed data by specifying this by using vcf@fix
```{r}
tail(vcf@fix)
```

#  There is a column in the fix portion of the data called QUAL. It is not defined in the meta portion of the data because it is defined in the VCF specification. It stands for ‘quality’. Does QUAL appear useful to us? Why or why not?

# High QUAL scores indicate high confidence calls. Although traditionally people use integer phred scores, this field is permitted to be a floating point to enable higher resolution for low confidence calls if desired. It is an alternate way too look at the quality of the frequencies. 

```{r}
plot(vcf)
```

# 5.  How would we query the sample names?

# The sample names are the genotypes or gt in the VFC. These are listed as the column names. To get the names of the colums use the function caolnames(). Specify that we need gt using @gt.

```{r}
colnames(vcf@gt)
```

# Part II - Analysis of Genome Data
```{r}
library('vcfR')
vcf <- read.vcfR("myVCFdata_filtered.vcf.gz")
```
```{r}
vcf
```
```{r}
x <- vcfR2genlight(vcf)
```
```{r}
x
```
```{r}
gt <- extract.gt(vcf, element = "GT")
gt[c(2,6,18), 1:3]
```
```{r}
t(as.matrix(x))[c(1,5,17), 1:3]
```

```{r}
library(adegenet)
```
```{r}
pop(x) <- as.factor(c("us", "eu", "us", "af", "eu", "us", "mx", "eu", "eu", "sa", "mx", "sa", "us", "sa", "Pmir", "us", "eu", "eu"))
popNames(x)
```

```{r}
ploidy(x) <- 2
```

```{r}
x.dist <- dist(x)
```

```{r}
x.dist <- poppr::bitwise.dist(x)
```
# Creating chromR Objects
```{r}
library(vcfR)

# Find the files.
vcf_file <- system.file("extdata", "pinf_sc50.vcf.gz", package = "pinfsc50")
dna_file <- system.file("extdata", "pinf_sc50.fasta", package = "pinfsc50")
gff_file <- system.file("extdata", "pinf_sc50.gff", package = "pinfsc50")

# Input the files.
vcf <- read.vcfR(vcf_file, verbose = FALSE)
dna <- ape::read.dna(dna_file, format = "fasta")
gff <- read.table(gff_file, sep="\t", quote="")

# Create a chromR object.
chrom <- create.chromR(name="Supercontig", vcf=vcf, seq=dna, ann=gff, verbose=TRUE)
```
```{r}
chrom
```

```{r}
plot(chrom)
```
```{r}
chromoqc(chrom, dp.alpha = 66)
```
# Process chrom data using proc.chromR(data, verbose = TRUE)
```{r}
chrom <- proc.chromR(chrom, verbose = TRUE)
```

```{r}
plot(chrom)
```
```{r}
chromoqc(chrom, dp.alpha = 66)
```
```{r}
chrom <- create.chromR(name="Supercontig", vcf=vcf, seq=dna, ann=gff, verbose=FALSE)
chrom <- proc.chromR(chrom, verbose = FALSE)
chromoqc(chrom, dp.alpha = 66)
```
```{r}
head(chrom@var.info)
```
```{r}
head(chrom@win.info)
```
# Genetic Differentation
#To address the question of within population diversity geneticists typically report heterozygosity. This is the probability that two alleles randomly chosen from a population will be different (Nei 1973). Ecologists may know this as Simpson’s Index (Simpson 1949). To address differentiation population geneticists typically utilize FST or one of its analogues. 
```{r}
library(vcfR)
data(vcfR_example)
pop <- as.factor(c("us", "eu", "us", "af", "eu", "us", "mx", "eu", "eu", "sa", "mx", "sa", "us", "sa", "Pmir", "us", "eu", "eu"))
myDiff <- genetic_diff(vcf, pops = pop, method = 'nei')
knitr::kable(head(myDiff[,1:15]))
```


```{r}
knitr::kable(head(myDiff[,16:19]))
```

# Summarize information, bring together columns or use violin plots
```{r}
knitr::kable(round(colMeans(myDiff[,c(3:9,16,19)], na.rm = TRUE), digits = 3))
```

```{r}
library(reshape2)
library(ggplot2)

dpf <- melt(myDiff[,c(3:8,19)], varnames=c('Index', 'Sample'), value.name = 'Depth', na.rm=TRUE)

```

```{r}
p <- ggplot(dpf, aes(x=variable, y=Depth)) + geom_violin(fill="#2ca25f", adjust = 1.2)
p <- p + xlab("")
p <- p + ylab("")
p <- p + theme_bw()
p
```
#Exercises Part II
# 1. You actually have everything you need to make a Manhattan plot. Can you figure out how to plot G′ST (y-axis) by genomic position (POS)?

# You would need the function getPOS(), then fill in the information for the structure of a Manhattan plot to show significance as -log(p-value) on y axis against chromosomes on x axis. Should be able to see regions of the genome that are signficant over many variants.
```{r}
plot(getPOS(vcf), myDiff$Gprimest,  pch = 20, col = "#1E90FF44", xlab = "", ylab = "", ylim = c(0, 1), xaxt = "n")
axis(side = 1, at = seq(0, 1e5, by = 1e4), labels = seq(0, 100, by = 10))
title(xlab='Genomic position (Kbp)')
title(ylab = expression(italic("G'"["ST"])))
```
# 2. This Manhatttan plot shouldlook a bit unusual. Can you think of anything that may be wrong with this analysis?

# Considering this plot is for many varinats at once, these sample sizes of variants are too small. This can be shown by looking at the pop table.
```{r}
table(pop)
```
# 3. Can you figure out how to zoom in on a particular region of a chromosome in chromoqc()?

```{r}
chromoqc(chrom, dp.alpha = 66, xlim = c(2e05, 4e05))
```
# 4. Can you use the function queryMETA() to look for other data in your file that may be of interest?

# Look at all the meta data using this function on vcf.
```{r}
queryMETA(vcf)
```

