---
title: "Lab8"
author: "Andrea Dame"
date: "10/19/2020"
output: html_document
---
# Exercise 1
# NGPhylogeny.fr is a web server that provides a basic workflow of sequence alignment, phylogenetic analysis, and tree visualization. This server is good for visualizing relationships between a small set of sequences. 
# Here a small set of core DNA sequences of small subunit ribosomal gene is used. The sequences are aligned so that a distance can be calculated. This is used to determine relationships between the sequences. These relationships can then be visualized through trees. A OneClick workflow was used to provide the tree below. 

```{r}
ppi <- 300
png("tree_00.png", width=3*ppi, height=3*ppi, res=ppi)
```

![Tree from NGPhylogeny.fr](tree_00.png)


#Exercise 2
# Tree file of core DNA sequences of small subunit ribosomal gene from CIPRES. The fastree_result.tre was downloaded to this computer. 


#Exercise 3
# Visualizing and Annotating Phylogenetic Trees with R+ggtree

```{r}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
```

```{r}
library(tidyverse)
library(ggtree)
```

```{r}
library(treeio)
library(ggimage)
```

```{r}
tree <- read.tree("tree_newick.nwk")
tree
```

```{r}
ggplot(tree) + geom_tree() + theme_tree()
ggtree(tree)
```

```{r}
ggtree(tree) + geom_treescale()
ggtree(tree) + theme_tree2()
```

```{r}
ggtree(tree, branch.length="none")
```

```{r}
ggtree(tree, branch.length="none", color="blue", size=2, linetype=3)
```

# Exercise 3.1 Make slanted, circular, and circular unscaled cladogram
```{r}
?ggtree

```

```{r}
ps <- ggtree(tree, layout="slanted") + ggtitle("slanted")
```

```{r}
pc <- ggtree(tree, layout="circular") + ggtitle("circular")
```

```{r}
pcuc <- ggtree(tree, layout="circular", branch.length="none", color="red", size=2, linetype=3) + ggtitle("(Cladogram) circular layout with Thick Red Lines")
```

```{r}
# create the basic plot
p <- ggtree(tree)
# add node points
p + geom_nodepoint()
# add tip points
p + geom_tippoint()
# Label the tips
p + geom_tiplab()
```

# Exercise 3.2 Create a phylogeny with the following aesthetic characteristics:
```{r}
p <- ggtree(tree)
p + geom_tippoint()
p + geom_nodepoint(aes(color = "yellow"))
p + geom_tiplab(aes(color = "purple", alpha = 0.5 ))
p + ggtitle("Aesthetic Phylogeny")
```


```{r}
ggtree(tree) + geom_text(aes(label=node), hjust=-.3)
```

```{r}
ggtree(tree) + geom_tiplab()
```

```{r}
MRCA(tree, c("C", "E"))
MRCA(tree, c("G", "H"))
```

```{r}
ggtree(tree) + 
  geom_cladelabel(node=17, label="Some random clade", color="red")
```

```{r}
ggtree(tree) + 
  geom_tiplab() + 
  geom_cladelabel(node=17, label="Some random clade", 
                  color="red2", offset=.8)
```

```{r}
ggtree(tree) + 
  geom_tiplab() + 
  geom_cladelabel(node=17, label="Some random clade", 
                  color="red2", offset=.8) + 
  geom_cladelabel(node=21, label="A different clade", 
                  color="blue", offset=.8)
```

```{r}
ggtree(tree) + 
  geom_tiplab() + 
  geom_cladelabel(node=17, label="Some random clade", 
                  color="red2", offset=.8, align=TRUE) + 
  geom_cladelabel(node=21, label="A different clade", 
                  color="blue", offset=.8, align=TRUE) + 
  theme_tree2() + 
  xlim(0, 70) + 
  theme_tree()
```

```{r}
ggtree(tree) + 
  geom_tiplab() + 
  geom_hilight(node=17, fill="gold") + 
  geom_hilight(node=21, fill="purple")
```

```{r}
ggtree(tree) + 
  geom_tiplab() + 
  geom_taxalink("E", "H", color="blue3") +
  geom_taxalink("C", "G", color="orange2", curvature=-.9)
```
# Exercise 3.3 Produce the figure
```{r}
MRCA(tree, tip=c("B", "C"))
MRCA(tree, tip=c("L", "J"))
ggtree(tree) + 
  geom_tiplab() + 
  geom_taxalink("C", "E", color="blue3") +
  geom_taxalink("G", "J", color="orange2", linetype=2, curvature=-.9) + geom_hilight(node=17, fill="gold") + 
  geom_hilight(node=21, fill="purple")
```

```{r}
# Read the data
tree <- read.beast("data/flu_tree_beast.tree")
# supply a most recent sampling date so you get the dates
# and add a scale bar
ggtree(tree, mrsd="2013-01-01") + 
  theme_tree2() 
# Finally, add tip labels and adjust axis
ggtree(tree, mrsd="2013-01-01") + 
  theme_tree2() + 
  geom_tiplab(align=TRUE, linesize=.5) + 
  xlim(1990, 2020)
```

```{r}
msaplot(p=ggtree(tree), fasta="data/flu_aasequence.fasta", window=c(150, 175))
```

```{r}
set.seed(42)
trees <- lapply(rep(c(10, 25, 50, 100), 3), rtree)
class(trees) <- "multiPhylo"
ggtree(trees) + facet_wrap(~.id, scale="free", ncol=4) + ggtitle("Many trees. Such phylogenetics. Wow.")
```

```{r}
# Generate a random tree with 30 tips
tree <- rtree(30)
# Make the original plot
p <- ggtree(tree)
# generate some random values for each tip label in the data
d1 <- data.frame(id=tree$tip.label, val=rnorm(30, sd=3))
# Make a second plot with the original, naming the new plot "dot", 
# using the data you just created, with a point geom.
p2 <- facet_plot(p, panel="dot", data=d1, geom=geom_point, aes(x=val), color='red3')
# Make some more data with another random value.
d2 <- data.frame(id=tree$tip.label, value = abs(rnorm(30, mean=100, sd=50)))
# Now add to that second plot, this time using the new d2 data above, 
# This time showing a bar segment, size 3, colored blue.
p3 <- facet_plot(p2, panel='bar', data=d2, geom=geom_segment, 
           aes(x=0, xend=value, y=y, yend=y), size=3, color='blue4') 
# Show all three plots with a scale
p3 + theme_tree2()
```

```{r}
# get phylopic 

newick <- "((Pongo_abelii,(Gorilla_gorilla_gorilla,(Pan_paniscus,Pan_troglodytes)Pan,Homo_sapiens)Homininae)Hominidae,Nomascus_leucogenys)Hominoidea;"

tree <- read.tree(text=newick)

d <- ggimage::phylopic_uid(tree$tip.label)
d$body_mass = c(52, 114, 47, 45, 58, 6)

p <- ggtree(tree) %<+% d + 
  geom_tiplab(aes(image=uid, colour=body_mass), geom="phylopic", offset=2.5) +
  geom_tiplab(aes(label=label), offset = .2) + xlim(NA, 7) +
  scale_color_viridis_c()
p  
```

# Exercise 4
# Tree file from the FastTreeMP output on CIPRES 
```{r}
tree2 <- read.tree("fastree_result.tre")
```

```{r}
ggtree(tree2, aes(color=group), layout="circular") + geom_tiplab() +
ggtitle("circular")
```



