---
title: "RNAseqLab7"
author: "Andrea Dame"
date: "10/12/2020"
output: html_document
---
# R Script
## ----style, echo=FALSE, message=FALSE, warning=FALSE, results="asis"----------
#load libraries
```{r}
library("BiocStyle")
library("knitr")
library("rmarkdown")
opts_chunk$set(message = FALSE, error = FALSE, warning = FALSE,
               cache = FALSE, fig.width = 5, fig.height = 5)
```

## ----load airway library for this data tutorial-------------------------
```{r}
library("airway")
```

## ---- set directory----------------------------------------------------------------------
```{r}
dir <- system.file("extdata", package="airway", mustWork=TRUE)
```

## ----list files in directory---------------------------------------------------------------
```{r}
list.files(dir)
list.files(file.path(dir, "quants"))
```

## ----First look at the data---------------------------------------------------------------
```{r}
csvfile <- file.path(dir, "sample_table.csv")
coldata <- read.csv(csvfile, row.names=1, stringsAsFactors=FALSE)
coldata
```


## ----make data frame  of metadata------------------------------------------------------
```{r}
coldata <- coldata[1:2,]
coldata$names <- coldata$Run
coldata$files <- file.path(dir, "quants", coldata$names, "quant.sf.gz")
file.exists(coldata$files)
```


## ---- libarary tximeta package for coldata, which is for transciript quantification import, which will locate and download the relevant annotation data from various sources, message=TRUE------------------------------
```{r}
library("tximeta")
se <- tximeta(coldata)
```


## ----look at the head of object se data frame----------------------------------------------------
```{r}
dim(se)
head(rownames(se))
```


## ----summarize to gene for object se,which summarizes the transcript-level quantifications to the gene level, message=TRUE-----------------------------------------
```{r}
gse <- summarizeToGene(se)
```


## ----look at ththe dimensions and head of new object gse------------------------------------------------
```{r}
dim(gse)
head(rownames(gse))
```


## ----sumexp, which is the components and parts of the experiment overall, echo=FALSE-------------------------------------------------------
```{r}
par(mar=c(0,0,0,0))
plot(1,1,xlim=c(0,100),ylim=c(0,100),bty="n",
     type="n",xlab="",ylab="",xaxt="n",yaxt="n")
polygon(c(45,90,90,45),c(5,5,70,70),col="pink",border=NA)
polygon(c(45,90,90,45),c(68,68,70,70),col="pink3",border=NA)
text(67.5,40,"assay(s)")
text(67.5,35,'e.g. "counts", ...')
polygon(c(10,40,40,10),c(5,5,70,70),col="skyblue",border=NA)
polygon(c(10,40,40,10),c(68,68,70,70),col="skyblue3",border=NA)
text(25,40,"rowRanges")
polygon(c(45,90,90,45),c(75,75,95,95),col="palegreen",border=NA)
polygon(c(45,47,47,45),c(75,75,95,95),col="palegreen3",border=NA)
text(67.5,85,"colData")
```

## ----loadfullgse, this loads information about different aspects on the dataframe, including the assays --------------------------------------------------------------
```{r}
data(gse)
gse
```

## ----assaysgse, so now we just want to look at the assays, which are count, abundance, and length----------------------------------------------------------------
```{r}
assayNames(gse)
head(assay(gse), 3)
colSums(assay(gse))
```


## ----rowrangesgse, this will show the ranges for the first five and last five genes------------------------------------------
```{r}
rowRanges(gse)
```


## ----lookseqinfo, shows more metadata about the sequences (chromosomes)------------------------------------
```{r}
seqinfo(rowRanges(gse))
```

## ----coldatagse,this is the colData for the summarized experiment which reflects the data.frame that was provided to the tximeta function for importing the quantification data. This will show the name, donor, and condition or treatment---------------------------------------------------------------
```{r}
colData(gse)
```

## ----gsevars, looking at the columns of just donor and condition------------------------------------------------------------------
```{r}
gse$donor
gse$condition
```

## ----gsevarsrename, renaming the variables from donor to cell, and condition to dex, which refers to the treatment in this experiment ------------------------------------------------------------
```{r}
gse$cell <- gse$donor
gse$dex <- gse$condition
```

## ----rename levels, changing the names from Untreated to untrt and Dexamethasone to trt, the order in which they appear in the data set must stay the same during this renaming -------------------------------------------------------------
```{r}
levels(gse$dex)
# when renaming levels, the order must be preserved!
levels(gse$dex) <- c("untrt", "trt")
```


## ----gsedex, library magrittr-, the relevel function decides how the variables will be coded, and how contrasts will be computed------------------------------------------------------------------
```{r}
library("magrittr")
gse$dex %<>% relevel("untrt")
gse$dex
```

## ----explaincmpass, eval = FALSE----------------------------------------------
#  gse$dex <- relevel(gse$dex, "untrt")

## ----countreads, quick check for fragments that could be mapped, the numbers tell how many decimal points to keep ---------------------------------------------------------------
```{r}
round( colSums(assay(gse)) / 1e6, 1 )
```

## ----loaddeseq2, loading the analysis program---------------------------------------------------------------
```{r}
library("DESeq2")
```

## ----makedds, add appropriate design for this analysis------------------------------------------------------------------
```{r}
dds <- DESeqDataSet(gse, design = ~ cell + dex)
```

## creating object countdata which makes count matrix of actual data, shows the first few lines using the heqad function-----------------------------------------------------------------------------
```{r}
countdata <- round(assays(gse)[["counts"]])
head(countdata, 3)
```

## Important to check what data looks like-----------------------------------------------------------------------------
```{r}
coldata <- colData(gse)
```

## To construct the DESeqDataSet object from the matrix of counts and the sample information table-----------------------------------------------------------------------------
```{r}
ddsMat <- DESeqDataSetFromMatrix(countData = countdata,
                                 colData = coldata,
                                 design = ~ cell + dex)

```

## removing the rows of the DESeq data set that have no counts, or only a single count across all samples, this increases speed of functions later on -----------------------------------------------

```{r}
nrow(dds)
keep <- rowSums(counts(dds)) > 1
dds <- dds[keep,]
nrow(dds)
```

## specifying that at least 3 samples have a count of 10 or higher-----------------------------------------------------------------------------
```{r}
keep <- rowSums(counts(dds) >= 10) >= 3
```


## ----meanSdCts, exploratory analysis for multidemensional data,to show this property of counts with some simulated data (here, Poisson counts with a range of lambda from 0.1 to 100). We plot the standard deviation of each row (genes) against the mean ----------------------------------------------
```{r}
lambda <- 10^seq(from = -1, to = 2, length = 1000)
cts <- matrix(rpois(1000*100, lambda), ncol = 100)
library("vsn")
meanSdPlot(cts, ranks = FALSE)
```


## ----meanSdLogCts, log transformed counts------------------------------------------------------------
```{r}
log.cts.one <- log2(cts + 1)
meanSdPlot(log.cts.one, ranks = FALSE)
```

## ----vst variance stabilizing transformation ----------------------------------------------------------------------
```{r}
vsd <- vst(dds, blind = FALSE)
head(assay(vsd), 3)
colData(vsd)
```

## ----rlog regularized logarithm transformation, which to choose? the VST is much faster to compute and is less sensitive to high count outliers than the rlog. The rlog tends to work well on small datasets (n < 30)---------------------------------------------------------------------

```{r}
rld <- rlog(dds, blind = FALSE)
head(assay(rld), 3)
```

## ----transformplot, fig.width = 6, fig.height = 2.5, to show the effect of transformation, plotting the first sample against the second, first simply using the log2 function (after adding 1, to avoid taking the log of zero), and then using the VST and rlog-transformed values---------------------------
```{r}
library("dplyr")
library("ggplot2")

dds <- estimateSizeFactors(dds)

df <- bind_rows(
  as_data_frame(log2(counts(dds, normalized=TRUE)[, 1:2]+1)) %>%
    mutate(transformation = "log2(x + 1)"),
  as_data_frame(assay(vsd)[, 1:2]) %>% mutate(transformation = "vst"),
  as_data_frame(assay(rld)[, 1:2]) %>% mutate(transformation = "rlog"))

colnames(df)[1:2] <- c("x", "y")  

lvls <- c("log2(x + 1)", "vst", "rlog")
df$transformation <- factor(df$transformation, levels=lvls)

ggplot(df, aes(x = x, y = y)) + geom_hex(bins = 80) +
  coord_fixed() + facet_grid( . ~ transformation)  
```


## dist function is used to calculate the Euclidean distance between samples. Use this on VST data.  data-----------------------------------------------------------------------------
```{r}
sampleDists <- dist(t(assay(vsd)))
sampleDists
```

## visualizing these distances in a heatmap, using Colorbrewer to add color-----------------------------------------------------------------------------

```{r}
library("pheatmap")
library("RColorBrewer")
```

## ----distheatmap, fig.width = 6.1, fig.height = 4.5, plot the sample distance matrix with the rows/columns arranged by the distances---------------------------
```{r}
sampleDistMatrix <- as.matrix( sampleDists )
rownames(sampleDistMatrix) <- paste( vsd$dex, vsd$cell, sep = " - " )
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         col = colors)

```

## Another package used to calculate sample distances-----------------------------------------------------------------------------
```{r}
library("PoiClaClu")
poisd <- PoissonDistance(t(counts(dds)))
```


## ----poisdistheatmap, fig.width = 6.1, fig.height = 4.5, then this is turned into a heatmap-----------------------
```{r}
samplePoisDistMatrix <- as.matrix( poisd$dd )
rownames(samplePoisDistMatrix) <- paste( dds$dex, dds$cell, sep=" - " )
colnames(samplePoisDistMatrix) <- NULL
pheatmap(samplePoisDistMatrix,
         clustering_distance_rows = poisd$dd,
         clustering_distance_cols = poisd$dd,
         col = colors)
```

## ----plotpca, fig.width=6, fig.height=4.5, a PCA plot or principlal components analysis-------------------------------------
```{r}
plotPCA(vsd, intgroup = c("dex", "cell"))
```


## A PCA plot using the VST data-----------------------------------------------------------------------------
```{r}
pcaData <- plotPCA(vsd, intgroup = c( "dex", "cell"), returnData = TRUE)
pcaData
percentVar <- round(100 * attr(pcaData, "percentVar"))
```


## ----ggplotpca, fig.width=6, fig.height=4.5, specifying the the color of the points which reflects dexamethasone treatment and the shape whichs reflects the cell line-----------------------------------
```{r}
ggplot(pcaData, aes(x = PC1, y = PC2, color = dex, shape = cell)) +
  geom_point(size =3) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() +
  ggtitle("PCA with VST data")
```


## Package for generalized principal component analysis which is another technique for performing dimension reduction on data that is not normally distributed----------------------------------------------------------------------------
```{r}
library("glmpca")
gpca <- glmpca(counts(dds), L=2)
gpca.dat <- gpca$factors
gpca.dat$dex <- dds$dex
gpca.dat$cell <- dds$cell
```


## ----glmpca, fig.width=6, fig.height=4.5, plotting this generalized PCA--------------------------------------
```{r}
ggplot(gpca.dat, aes(x = dim1, y = dim2, color = dex, shape = cell)) +
  geom_point(size =3) + coord_fixed() + ggtitle("glmpca - Generalized PCA")
```

## ----mdsvst, fig.width=6, fig.height=4.5, multidimensional scaling, similar to PCA--------------------------------------
```{r}
mds <- as.data.frame(colData(vsd))  %>%
  cbind(cmdscale(sampleDistMatrix))
ggplot(mds, aes(x = `1`, y = `2`, color = dex, shape = cell)) +
  geom_point(size = 3) + coord_fixed() + ggtitle("MDS with VST data")
```

## ----mdspois, fig.width=6, fig.height=4.5, this is the same plot for the poisson distance-------------------------------------
```{r}
mdsPois <- as.data.frame(colData(dds)) %>%
  cbind(cmdscale(samplePoisDistMatrix))
ggplot(mdsPois, aes(x = `1`, y = `2`, color = dex, shape = cell)) +
  geom_point(size = 3) + coord_fixed() + ggtitle("MDS with PoissonDistances")
```

## ----airwayDE, run the differential expression pipeline on the raw counts with this function-----------------------------------------------------------------
```{r}
dds <- DESeq(dds)
```


## extract the estimated log2 fold changes and p values for the last variable in the design formula by calling results without argument-----------------------------------------------------------------------------
```{r}
res <- results(dds)
res
```
## other types of comparison
## Adding contrast argument to results-----------------------------------------------------------------------------
```{r}
res <- results(dds, contrast=c("dex","trt","untrt"))
```


## -----------------------------------------------------------------------------
```{r}
mcols(res, use.names = TRUE)
```

## -----------------------------------------------------------------------------
```{r}
summary(res)
```


## -----------------------------------------------------------------------------
```{r}
res.05 <- results(dds, alpha = 0.05)
table(res.05$padj < 0.05)

```

## -----------------------------------------------------------------------------

```{r}
resLFC1 <- results(dds, lfcThreshold=1)
table(resLFC1$padj < 0.1)
```

## -----------------------------------------------------------------------------
```{r}
results(dds, contrast = c("cell", "N061011", "N61311"))
```

## Multiple testing to be sure not to use p value as direct evidence against the null
## ----sumres-------------------------------------------------------------------
```{r}
sum(res$pvalue < 0.05, na.rm=TRUE)
sum(!is.na(res$pvalue))
```


## -----------------------------------------------------------------------------
```{r}
sum(res$padj < 0.1, na.rm=TRUE)
```


## -----------------------------------------------------------------------------
```{r}
resSig <- subset(res, padj < 0.1)
head(resSig[ order(resSig$log2FoldChange), ])
```


## -----------------------------------------------------------------------------

```{r}
head(resSig[ order(resSig$log2FoldChange, decreasing = TRUE), ])
```
##Here we start plotting the results
## ----plotcounts---------------------------------------------------------------
```{r}
topGene <- rownames(res)[which.min(res$padj)]
plotCounts(dds, gene = topGene, intgroup=c("dex"))
```


## ----ggplotcountsjitter, fig.width = 4, fig.height = 3------------------------
```{r}
library("ggbeeswarm")
geneCounts <- plotCounts(dds, gene = topGene, intgroup = c("dex","cell"),
                         returnData = TRUE)
ggplot(geneCounts, aes(x = dex, y = count, color = cell)) +
  scale_y_log10() +  geom_beeswarm(cex = 3)
```


## ----ggplotcountsgroup, fig.width = 4, fig.height = 3-------------------------
```{r}
ggplot(geneCounts, aes(x = dex, y = count, color = cell, group = cell)) +
  scale_y_log10() + geom_point(size = 3) + geom_line()
```

## MA- plot, the y-axis, the “M” stands for “minus” – subtraction of log values is equivalent to the log of the ratio – and on the x-axis, the “A” stands for “average"
## ----plotma-------------------------------------------------------------------
```{r}
library("apeglm")
resultsNames(dds)
res <- lfcShrink(dds, coef="dex_trt_vs_untrt", type="apeglm")
plotMA(res, ylim = c(-5, 5))
```


## ----plotmaNoShr--------------------------------------------------------------
```{r}
res.noshr <- results(dds, name="dex_trt_vs_untrt")
plotMA(res.noshr, ylim = c(-5, 5))
```


## ----plotmalabel--------------------------------------------------------------
```{r}
plotMA(res, ylim = c(-5,5))
topGene <- rownames(res)[which.min(res$padj)]
with(res[topGene, ], {
  points(baseMean, log2FoldChange, col="dodgerblue", cex=2, lwd=2)
  text(baseMean, log2FoldChange, topGene, pos=2, col="dodgerblue")
})
```


## ----histpvalue2--------------------------------------------------------------
```{r}
hist(res$pvalue[res$baseMean > 1], breaks = 0:20/20,
     col = "grey50", border = "white")
```

## Gene clustering, as visualized by the heat maps
## -----------------------------------------------------------------------------
```{r}
library("genefilter")
topVarGenes <- head(order(rowVars(assay(vsd)), decreasing = TRUE), 20)
```

##Heatmap plot of relative VST-transformed values across samples 
## ----genescluster-------------------------------------------------------------
```{r}
mat  <- assay(vsd)[ topVarGenes, ]
mat  <- mat - rowMeans(mat)
anno <- as.data.frame(colData(vsd)[, c("cell","dex")])
pheatmap(mat, annotation_col = anno)
```

## Independent filtering, The ratio of small p values for genes binned by mean normalized count
## ----sensitivityovermean, fig.width=6-----------------------------------------
```{r}
qs <- c(0, quantile(resLFC1$baseMean[resLFC1$baseMean > 0], 0:6/6))
bins <- cut(resLFC1$baseMean, qs)
levels(bins) <- paste0("~", round(signif((qs[-1] + qs[-length(qs)])/2, 2)))
fractionSig <- tapply(resLFC1$pvalue, bins, function(p)
  mean(p < .05, na.rm = TRUE))
barplot(fractionSig, xlab = "mean normalized count",
        ylab = "fraction of small p values")
```


## ---- eval=FALSE--------------------------------------------------------------
#Independent Hypothesis Weighting, to weight the hypotheses.

#Annotating and exporting results, helps with mapping various ID schemes to each other
## -----------------------------------------------------------------------------
```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")
```


## -----------------------------------------------------------------------------
```{r}
columns(org.Hs.eg.db)
```


## -----------------------------------------------------------------------------
```{r}
ens.str <- substr(rownames(res), 1, 15)
res$symbol <- mapIds(org.Hs.eg.db,
                     keys=ens.str,
                     column="SYMBOL",
                     keytype="ENSEMBL",
                     multiVals="first")
res$entrez <- mapIds(org.Hs.eg.db,
                     keys=ens.str,
                     column="ENTREZID",
                     keytype="ENSEMBL",
                     multiVals="first")
```


## Exporting these results-----------------------------------------------------------------------------
```{r}
resOrdered <- res[order(res$pvalue),]
head(resOrdered)
```


## ----eval=FALSE---------------------------------------------------------------
#Plotting fold changes in genomic space, plot differential expression results in genomic space
## -----------------------------------------------------------------------------
```{r}
resGR <- lfcShrink(dds, coef="dex_trt_vs_untrt", type="apeglm", format="GRanges")
resGR
```


## -----------------------------------------------------------------------------
```{r}
ens.str <- substr(names(resGR), 1, 15)
resGR$symbol <- mapIds(org.Hs.eg.db, ens.str, "SYMBOL", "ENSEMBL")
```


## -----------------------------------------------------------------------------
```{r}
library("Gviz")
```

## -----------------------------------------------------------------------------
```{r}
window <- resGR[topGene] + 1e6
strand(window) <- "*"
resGRsub <- resGR[resGR %over% window]
naOrDup <- is.na(resGRsub$symbol) | duplicated(resGRsub$symbol)
resGRsub$group <- ifelse(naOrDup, names(resGRsub), resGRsub$symbol)
```


## -----------------------------------------------------------------------------
```{r}
status <- factor(ifelse(resGRsub$padj < 0.05 & !is.na(resGRsub$padj),
                        "sig", "notsig"))
```


## ----gvizplot-----------------------------------------------------------------
```{r}
options(ucscChromosomeNames = FALSE)
g <- GenomeAxisTrack()
a <- AnnotationTrack(resGRsub, name = "gene ranges", feature = status)
d <- DataTrack(resGRsub, data = "log2FoldChange", baseline = 0,
               type = "h", name = "log2 fold change", strand = "+")
plotTracks(list(g, d, a), groupAnnotation = "group",
           notsig = "grey", sig = "hotpink")
```

# Removing hidden batch effects, sva provides statistical methods
## -----------------------------------------------------------------------------
```{r}
library("sva")
```
# surrogate variables 1 and 2 plotted over cell line
## -----------------------------------------------------------------------------
```{r}
dat  <- counts(dds, normalized = TRUE)
idx  <- rowMeans(dat) > 1
dat  <- dat[idx, ]
mod  <- model.matrix(~ dex, colData(dds))
mod0 <- model.matrix(~   1, colData(dds))
svseq <- svaseq(dat, mod, mod0, n.sv = 2)
svseq$sv
```

## ----svaplot------------------------------------------------------------------
```{r}
par(mfrow = c(2, 1), mar = c(3,5,3,1))
for (i in 1:2) {
  stripchart(svseq$sv[, i] ~ dds$cell, vertical = TRUE, main = paste0("SV", i))
  abline(h = 0)
}
```


## -----------------------------------------------------------------------------
```{r}
ddssva <- dds
ddssva$SV1 <- svseq$sv[,1]
ddssva$SV2 <- svseq$sv[,2]
design(ddssva) <- ~ SV1 + SV2 + dex
```

#RUVseq function is used to estimate factors of unwanted variation, analogous to SVA’s surrogate variables
## -----------------------------------------------------------------------------
```{r}
library("RUVSeq")
```


## -----------------------------------------------------------------------------
```{r}
set <- newSeqExpressionSet(counts(dds))
idx  <- rowSums(counts(set) > 5) >= 2
set  <- set[idx, ]
set <- betweenLaneNormalization(set, which="upper")
not.sig <- rownames(res)[which(res$pvalue > .1)]
empirical <- rownames(set)[ rownames(set) %in% not.sig ]
set <- RUVg(set, empirical, k=2)
pData(set)
```


## ----ruvplot------------------------------------------------------------------
```{r}
par(mfrow = c(2, 1), mar = c(3,5,3,1))
for (i in 1:2) {
  stripchart(pData(set)[, i] ~ dds$cell, vertical = TRUE, main = paste0("W", i))
  abline(h = 0)
}
```


## -----------------------------------------------------------------------------
```{r}
ddsruv <- dds
ddsruv$W1 <- set$W_1
ddsruv$W2 <- set$W_2
design(ddsruv) <- ~ W1 + W2 + dex
```

#Time course experiments, such as finding genes that react in a condition-specific manner over time
## -----------------------------------------------------------------------------
```{r}
library("fission")
data("fission")
ddsTC <- DESeqDataSet(fission, ~ strain + minute + strain:minute)
```

## ----fissionDE----------------------------------------------------------------
```{r}
ddsTC <- DESeq(ddsTC, test="LRT", reduced = ~ strain + minute)
resTC <- results(ddsTC)
resTC$symbol <- mcols(ddsTC)$symbol
head(resTC[order(resTC$padj),], 4)
```
#Normalized counts for a gene with condition-specific change over time
## ----fissioncounts, fig.width=6, fig.height=4.5-------------------------------
```{r}
fiss <- plotCounts(ddsTC, which.min(resTC$padj), 
                   intgroup = c("minute","strain"), returnData = TRUE)
fiss$minute <- as.numeric(as.character(fiss$minute))
ggplot(fiss,
       aes(x = minute, y = count, color = strain, group = strain)) + 
  geom_point() + stat_summary(fun.y=mean, geom="line") +
  scale_y_log10()
```


## -----------------------------------------------------------------------------
```{r}
resultsNames(ddsTC)
res30 <- results(ddsTC, name="strainmut.minute30", test="Wald")
res30[which.min(resTC$padj),]
```


## -----------------------------------------------------------------------------
```{r}
betas <- coef(ddsTC)
colnames(betas)
```

## ----fissionheatmap-----------------------------------------------------------
```{r}
topGenes <- head(order(resTC$padj),20)
mat <- betas[topGenes, -c(1,2)]
thr <- 3 
mat[mat < -thr] <- -thr
mat[mat > thr] <- thr
pheatmap(mat, breaks=seq(from=-thr, to=thr, length=101),
         cluster_col=FALSE)
```

# Reporting session information
## -----------------------------------------------------------------------------
```{r}
sessionInfo()
```


